components/dnd_custom_time_picker_modal/dnd_custom_time_picker_modal.tsx

```diff
@@ -0,0 +1,219 @@
+// See LICENSE.txt for license information.
+
+import React from 'react';
+import {FormattedMessage} from 'react-intl';
+import DayPickerInput from 'react-day-picker/DayPickerInput';
+
+import {ActionFunc} from 'mattermost-redux/types/actions';
+import {UserStatus} from 'mattermost-redux/types/users';
+
+import GenericModal from 'components/generic_modal';
+
+import {UserStatuses} from 'utils/constants';
+import Menu from 'components/widgets/menu/menu';
+import MenuWrapper from 'components/widgets/menu/menu_wrapper';
+
+import './dnd_custom_time_picker_modal.scss';
+import {toUTCUnix} from 'utils/datetime';
+
+type Props = {
+    onHide: () => void;
+    userId: string;
+    currentDate: Date;
+    actions: {
+        setStatus: (status: UserStatus) => ActionFunc;
+    };
+};
+
+type State = {
+    userId: string;
+    currentDate: Date;
+    selectedDate: string;
+    selectedTime: string;
+    timeMenuList: string[];
+}
+
+export default class DndCustomTimePicker extends React.PureComponent<Props, State> {
+    constructor(props: Props) {
+        super(props);
+
+        this.state = {
+            userId: props.userId || '',
+            currentDate: this.props.currentDate,
+            selectedDate: this.formatDate(this.props.currentDate) || '',
+            selectedTime: '',
+            timeMenuList: [],
+        };
+    }
+
+    formatDate = (date: Date): string => {
+        const month = (date.getMonth() + 1).toString().padStart(2, '0');
+        const day = date.getDate().toString().padStart(2, '0');
+        const year = date.getFullYear().toString();
+        return [year, month, day].join('-');
+    }
+
+    getText = () => {
+        const modalHeaderText = (
+            <FormattedMessage
+                id='dnd_custom_time_picker_modal.defaultMsg'
+                defaultMessage='Disable notifications till'
+            />
+        );
+        const confirmButtonText = (
+            <FormattedMessage
+                id='dnd_custom_time_picker_modal.submitButton'
+                defaultMessage='Disable Notifications'
+            />
+        );
+
+        return {
+            modalHeaderText,
+            confirmButtonText,
+        };
+    }
+
+    handleConfirm = () => {
+        const hours = parseInt(this.state.selectedTime.split(':')[0], 10);
+        const minutes = parseInt(this.state.selectedTime.split(':')[1], 10);
+        const endTime = new Date(this.state.selectedDate);
+        endTime.setHours(hours, minutes);
+        this.props.actions.setStatus({
+            user_id: this.props.userId,
+            status: UserStatuses.DND,
+            dnd_end_time: toUTCUnix(endTime),
+            manual: true,
+            last_activity_at: toUTCUnix(this.state.currentDate),
+        });
+        this.props.onHide();
+    }
+
+    handleDaySelection = (day: Date) => {
+        const dString = this.formatDate(day);
+        this.setState({
+            selectedDate: dString,
+        }, this.setTimeMenuList);
```

Comment:
Why call `setTimeMenuList` after setting the state? That'll cause the component to render twice when the date changes. Similarly, it'll also render twice on mount where `setTimeMenuList` is also called.

Could we maybe change `setTimeMenuList` so that instead of calling `setState` itself, we instead rename it something like `makeTimeMenuList` and have it return an object? Then we could use it here like
```javascript
this.setState({
    selectedDate: dString,
    ...this.makeTimeMenuList(dString),
});
```

Similarly, instead of calling it in `componentDidMount` then, we could just use it in the constructor.

Reply:
Makes sense, thanks. I have made the changes.